### DISCUSSION TITLE
Grain AST

### DISCUSSION THREAD

**Discussion by vereym** [grain-lang/grain#2115] [Help] [answered]
Hey,
I'm trying to understand how grain is structured and evaluated. Therefore, I wanted to know if there is a way to output the internal AST.
Could be also cool to feature grain on [astexplorer](https://astexplorer.net/)

**Comment by ospencer** [comment, marked as answer]
Hey @vereym! There isn't really a _clean_ way to do it because it's mostly for debugging, but if you compile with `--verbose` it'll dump all of the intermediate IRs, the first of which being the AST for the input program. It's quite a lot of output, so I'd recommend dumping it to a file. If you're willing to make some minor modifications to the compiler it's pretty easy to get it as JSON/write it to a file. I'll also mention that we have this doc (https://github.com/grain-lang/grain/blob/main/docs/contributor/compiler_walkthrough.md) which is a little outdated but explains the compilation process. There's an open PR to update it, though. https://github.com/grain-lang/grain/pull/2092

Have fun!
